### DISCUSSION TITLE
How to properly dump i/o messages and as4 receipts?

### DISCUSSION THREAD

**Discussion by iansmirlis** [phax/phase4#204] [Q&A] [answered]
Hello Philip

Sorry if this is already answered or if it should be obvious by the examples, but I need some verification I got it right
I would like to keep all the messages that pass through the access point in order to 

1) Debug if something goes wrong and
2) Prove to a 3rd party in the future that a message was actually receipt by the other AP

Regarding (1) I am implementing an IAS4ServletMessageProcessorSPI processor similar to your below example that saves dumps to disk
https://github.com/phax/phase4/blob/8c9fd26be5cf23813cdc4cb6e05ce9daceaba690/phase4-server-webapp/src/main/java/com/helger/phase4/server/spi/ExampleReceiveMessageProcessorSPI.java#L60

If I understand correctly, the processAS4ResponseMessage() dumps the outgoing AS4 responses FROM this AP (replies to other APs initiated requests). The processAS4UserMessage() and processAS4SignalMessage() dumps incoming AS4 messages and I will keep them all like your examples.

Regarding (2), I implemented a AbstractAS4RawResponseConsumer<> to use with msgbuilder.rawResponseConsumer(). I assume that on a succesful transmission, the AS4 response that reaches handleResponse() can be used as a message receipt.

Do you see anything wrong in the above reasonings?

By the way regarding sending files, I see that in the as4 response I get, I see:
```
<eb:MessageId>9d3478ca-377b-4168-8fa3-827af09db1ba@phase4</eb:MessageId>
<eb:RefToMessageId>54f6746e-033d-4a09-bee4-edcc3d182747@phase4</eb:RefToMessageId>
```      
Initially I thought that RefToMessageId would be the InstanceIdentifier, but it seems that MessageId is another UUID generated later by phase4. 

Do I have access to this during sending? Is there any processor similar to IAS4ServletMessageProcessorSPI in order to dump exported messages and final generated soap?

Many thanks for your help

**Comment by phax** [comment, marked as answer]
Well, for that reason I introduced the `IAS4IncomingDumper` and `IAS4OutgoingDumper` that give you the possibility to dump all messages on the wire, even the ones that are invalid AS4 messages. The SPI processors however are only invoked for incoming "valid AS4 messages". Global dumpers are set via `AS4DumpManager.setIncomingDumper` and `AS4DumpManager.setOutgoingDumper`. For sending you can also use the global once or use the builder methods `.incomingDumper` and `.outgoingDumper` to set specific instances per transmission.

This particular example `ExampleReceiveMessageProcessorSPI ` you mention has two original intentions: a) show that multiple SPI handlers are possible and b) show how to "de-onionize" the Peppol message by dumping the different levels. It's not meant for dumping the whole message.

And yes, your assumption that "AS4 Message ID" and "SBDH Instance ID" are two different things is correct. This is because a) the SBDH could be created by C1 already and b) in case of retries on an AS4 level a new AS4 message ID created foir each transmission. To access the AS4 message ID you can a) create a new UUID and pass it into the builder via `.messageID` or b) retrieve the internally created AS4 message as shown in the example in https://github.com/phax/phase4-peppol-standalone/blob/main/src/main/java/com/helger/phase4/peppolstandalone/controller/PeppolSenderController.java#L137

**Comment by iansmirlis** [comment]
Thanks for the guidance. One more question:

IAS4IncomingDumper and IAS4OutgoingDumper are very convenient when you are about to write to a file, but it gets tricky when you want to dump to a database. I.e. the flow in this case is to create a stream in memory in onNewRequest() but you will actually use it inside onEndRequest(), as it will be at this point that you have the completed stream to dump. (same as onBeginRequest() and onEndRequest())

The dumper implementation in this case has to somehow create streams in onNewRequest() and understand which one is completed in onEndRequest() in order to dump it to the database properly and discard it.

However there are 2 problems with this: a) The stream is closed by the caller of onNewRequest(), so I am not sure if it's safe to use it inside onEndRequest() (in C# closed memory streams are actually disposed) and b) It's not obvious how to map created and completed streams.

Maybe ideally I could have another callback that would give me the completed stream and some info about the message, before it's closed, i.e. onStreamClosing() or something like that.

Maybe I am missing something?

**Reply by phax** [reply]
I understand the issue. The classicial solution for this scenario is to use a temporary file. `onNewRequest` opens it, and returns a new `FileOutputStream` (or so). The mapping in `onEndRequest` depends on, whether it is a sent or a received message - that's what make it so tricky.

Per one outgoing AS4 message exchange:
1. Outgoing dumper: AS4 UserMessage - `eMsgMode = REQUEST`
2. Incoming dumper: AS4 SignalMessage - `IAS4IncomingMessageMetadata.getMode() = RESPONSE`
==> Correlate via "AS4 Message ID"
    * Outgoing dumper: `sMessageID`
    * Incoming dumper: `IAS4IncomingMessageMetadata.getRequestMessageID ()`

Per one incoming AS4 message exchange:
1. Incoming dumper: AS4 UserMessage - `IAS4IncomingMessageMetadata.getMode() = REQUEST`
2. Outgoing dumper: AS4 SignalMessage - `eMsgMode = RESPONSE`
==> Correlate via "Incoming Unique ID"
    * Incoming dumper: `IAS4IncomingMessageMetadata.getIncomingUniqueID ()`
    * Outgoing dumper: `IAS4IncomingMessageMetadata.getIncomingUniqueID ()`
